### PR TITLE
create nnc listener concept and adapt existing poolmonitor and swift …

### PR DESCRIPTION
**Pull request by rbtr**
…service to it

Signed-off-by: Evan Baker <rbtr@users.noreply.github.com>

**Reason for Change**:

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

**Notes**:

### CODE REVIEW

File: cns/singletenantcontroller/conversion.go [hunk 1]
```diff
@@ -20,6 +23,42 @@ var (
...
+func SwiftNodeNetworkConfigListener(cnscli cnsClient) NodeNetworkConfigListenerFunc {
+	return func(nnc *v1alpha.NodeNetworkConfig) error {
+		// Create NC request and hand it off to CNS
+		ncRequest, err := CRDStatusToNCRequest(&nnc.Status)
+		if err != nil {
+			return errors.Wrap(err, "failed to convert NNC status to network container request")
+		}
+		responseCode := cnscli.CreateOrUpdateNetworkContainerInternal(&ncRequest)
+		err = restserver.ResponseCodeToError(responseCode)
+		if err != nil {
```
Comment (thatmattlong) [inline cns/singletenantcontroller/conversion.go:51]: can combine `if err := restserver.ResponseCode ...; err != nil`

File: cns/singletenantcontroller/reconciler.go [hunk 2]
```diff
@@ -35,19 +28,21 @@ type nncGetter interface {
...
 // Reconciler watches for CRD status changes
 type Reconciler struct {
-	cnscli             cnsClient
-	ipampoolmonitorcli ipamPoolMonitorClient
-	nnccli             nncGetter
-	started            chan interface{}
-	once               sync.Once
+	nncListeners []nodeNetworkConfigListener
+	nnccli       nncGetter
+	once         sync.Once
```
Comment (thatmattlong) [inline cns/singletenantcontroller/reconciler.go:33]: using this anywhere?

**Review by thatmattlong — APPROVED** [review]

File: cns/singletenantcontroller/conversion.go [hunk 3]
```diff
@@ -20,6 +23,42 @@ var (
...
 
+type cnsClient interface {
+	CreateOrUpdateNetworkContainerInternal(*cns.CreateNetworkContainerRequest) cnstypes.ResponseCode
+}
+
+var _ nodeNetworkConfigListener = (NodeNetworkConfigListenerFunc)(nil) //nolint:gocritic // clarity
+
+type NodeNetworkConfigListenerFunc func(*v1alpha.NodeNetworkConfig) error
+
+func (f NodeNetworkConfigListenerFunc) Update(nnc *v1alpha.NodeNetworkConfig) error {
```
[the remark this answers is not in the document]
Comment (rbtr) [reply cns/singletenantcontroller/conversion.go:34]: it is a little on the abstract side, but it is cleaner than needing to attach these funcs to a stub struct as methods when the right closure provides all the context we need. it's the same pattern as `http.Handler/http.HandlerFunc` 🙂

File: cns/singletenantcontroller/conversion.go (hunk 1 above)
Comment (rbtr) [reply cns/singletenantcontroller/conversion.go:51]: incorporated this in the next one

**Pull request merged**